File: solution/kotlin/day07/src/test/kotlin/PipelineTests.kt
import ci.Config
import ci.Emailer
import ci.Pipeline
import ci.Project.ProjectBuilder
import ci.TestStatus.*
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.mockk.Called
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify

class PipelineTests : StringSpec({
    lateinit var config: Config
    lateinit var emailer: Emailer
    lateinit var log: CapturingLogger

    beforeEach {
        config = mockk()
        emailer = mockk()
        log = CapturingLogger()

        every { emailer(any()) } returns Unit
    }

    "project with tests that deploys successfully with email notification" {
        every { config.sendEmailSummary() } returns true

        ProjectBuilder()
            .setTestStatus(PassingTests)
            .setDeploysSuccessfully(true)
            .build()
            .let {
                Pipeline(config, emailer, log).run(it)
            }

        log.lines shouldBe listOf(
            "INFO: Tests passed",
            "INFO: Deployment successful",
            "INFO: Sending email"
        )
        verify { emailer("Deployment completed successfully") }
    }

    "project with tests that deploys successfully without email notification" {
        every { config.sendEmailSummary() } returns false

        ProjectBuilder()
            .setTestStatus(PassingTests)
            .setDeploysSuccessfully(true)
            .build()
            .let {
                Pipeline(config, emailer, log).run(it)
            }

        log.lines shouldBe listOf(
            "INFO: Tests passed",
            "INFO: Deployment successful",
            "INFO: Email disabled"
        )
        verify { emailer wasNot Called }
    }

    "project without tests that deploys successfully with email notification" {
        every { config.sendEmailSummary() } returns true

        ProjectBuilder()
            .setTestStatus(NoTests)
            .setDeploysSuccessfully(true)
            .build()
            .let {
                Pipeline(config, emailer, log).run(it)
            }

        log.lines shouldBe listOf(
            "INFO: No tests",
            "INFO: Deployment successful",
            "INFO: Sending email"
        )
        verify { emailer("Deployment completed successfully") }
    }

    "project without tests that deploys successfully without email notification" {
        every { config.sendEmailSummary() } returns false

        ProjectBuilder()
            .setTestStatus(NoTests)
            .setDeploysSuccessfully(true)
            .build()
            .let {
                Pipeline(config, emailer, log).run(it)
            }

        log.lines shouldBe listOf(
            "INFO: No tests",
            "INFO: Deployment successful",
            "INFO: Email disabled"
        )
        verify { emailer wasNot Called }
    }

    "project with tests that fail with email notification" {
        every { config.sendEmailSummary() } returns true

        ProjectBuilder()
            .setTestStatus(FailingTests)
            .build()
            .let {
                Pipeline(config, emailer, log).run(it)
            }

        log.lines shouldBe listOf(
            "ERROR: Tests failed",
            "INFO: Sending email"
        )
        verify { emailer("Tests failed") }
    }

    "project with tests that fail without email notification" {
        every { config.sendEmailSummary() } returns false

        ProjectBuilder()
            .setTestStatus(FailingTests)
            .build()
            .let {
                Pipeline(config, emailer, log).run(it)
            }

        log.lines shouldBe listOf(
            "ERROR: Tests failed",
            "INFO: Email disabled"
        )
        verify { emailer wasNot Called }
    }

    "project with tests and failing build with email notification" {
        every { config.sendEmailSummary() } returns true

        ProjectBuilder()
            .setTestStatus(PassingTests)
            .setDeploysSuccessfully(false)
            .build()
            .let {
                Pipeline(config, emailer, log).run(it)
            }

        log.lines shouldBe listOf(
            "INFO: Tests passed",
            "ERROR: Deployment failed",
            "INFO: Sending email"
        )
        verify { emailer("Deployment failed") }
    }

    "project with tests and failing build without email notification" {
        every { config.sendEmailSummary() } returns false

        ProjectBuilder()
            .setTestStatus(PassingTests)
            .setDeploysSuccessfully(false)
            .build()
            .let {
                Pipeline(config, emailer, log).run(it)
            }

        log.lines shouldBe listOf(
            "INFO: Tests passed",
            "ERROR: Deployment failed",
            "INFO: Email disabled"
        )
        verify { emailer wasNot Called }
    }

    "project without tests and failing build with email notification" {
        every { config.sendEmailSummary() } returns true

        ProjectBuilder()
            .setTestStatus(NoTests)
            .setDeploysSuccessfully(false)
            .build()
            .let {
                Pipeline(config, emailer, log).run(it)
            }

        log.lines shouldBe listOf(
            "INFO: No tests",
            "ERROR: Deployment failed",
            "INFO: Sending email"
        )
        verify { emailer("Deployment failed") }
    }

    "project without tests and failing build without email notification" {
        every { config.sendEmailSummary() } returns false

        ProjectBuilder()
            .setTestStatus(NoTests)
            .setDeploysSuccessfully(false)
            .build()
            .let {
                Pipeline(config, emailer, log).run(it)
            }

        log.lines shouldBe listOf(
            "INFO: No tests",
            "ERROR: Deployment failed",
            "INFO: Email disabled"
        )
        verify { emailer wasNot Called }
    }
})

File: solution/c#/Day07/Day07.Tests/PipelineTests.cs
using Day07.CI;
using Day07.CI.Dependencies;
using Day07.Tests.Doubles;
using FluentAssertions;
using NSubstitute;
using Xunit;
using static Day07.CI.Dependencies.TestStatus;
using static NSubstitute.Arg;

namespace Day07.Tests
{
    public class PipelineTests
    {
        private readonly IConfig _config = Substitute.For<IConfig>();
        private readonly CapturingLogger _log = new();
        private readonly IEmailer _emailer = Substitute.For<IEmailer>();
        private readonly Pipeline _pipeline;

        public PipelineTests() => _pipeline = new Pipeline(_config, _emailer, _log);

        [Fact]
        public void Project_With_Tests_That_Deploys_Successfully_With_Email_Notification()
        {
            _config.SendEmailSummary().Returns(true);

            var project = Project.Builder()
                .With(PassingTests)
                .Deployed(true)
                .Build();

            _pipeline.Run(project);

            _log.LoggedLines
                .Should()
                .BeEquivalentTo("INFO: Tests passed",
                    "INFO: Deployment successful",
                    "INFO: Sending email");

            _emailer.Received(1).Send("Deployment completed successfully");
        }

        [Fact]
        public void Project_Without_Tests_That_Deploys_Successfully_With_Email_Notification()
        {
            _config.SendEmailSummary().Returns(true);

            var project = Project.Builder()
                .With(NoTests)
                .Deployed(true)
                .Build();

            _pipeline.Run(project);

            _log.LoggedLines
                .Should()
                .BeEquivalentTo("INFO: No tests",
                    "INFO: Deployment successful",
                    "INFO: Sending email");

            _emailer.Received(1).Send("Deployment completed successfully");
        }

        [Fact]
        public void Project_Without_Tests_That_Deploys_Successfully_Without_Email_Notification()
        {
            _config.SendEmailSummary().Returns(false);

            var project = Project.Builder()
                .With(NoTests)
                .Deployed(true)
                .Build();

            _pipeline.Run(project);

            _log.LoggedLines
                .Should()
                .BeEquivalentTo("INFO: No tests",
                    "INFO: Deployment successful",
                    "INFO: Email disabled");

            _emailer.DidNotReceive().Send(Any<string>());
        }

        [Fact]
        public void Project_With_Tests_That_Fail_With_Email_Notification()
        {
            _config.SendEmailSummary().Returns(true);

            var project = Project.Builder()
                .With(FailingTests)
                .Deployed(true)
                .Build();

            _pipeline.Run(project);

            _log.LoggedLines
                .Should()
                .BeEquivalentTo("ERROR: Tests failed",
                    "INFO: Sending email");

            _emailer.Received(1).Send("Tests failed");
        }

        [Fact]
        public void Project_With_Tests_That_Fail_Without_Email_Notification()
        {
            _config.SendEmailSummary().Returns(false);

            var project = Project.Builder()
                .With(FailingTests)
                .Build();

            _pipeline.Run(project);

            _log.LoggedLines
                .Should()
                .BeEquivalentTo("ERROR: Tests failed",
                    "INFO: Email disabled");

            _emailer.DidNotReceive().Send(Any<string>());
        }

        [Fact]
        public void Project_With_Tests_And_Failing_Build_With_Email_Notification()
        {
            _config.SendEmailSummary().Returns(true);

            var project = Project.Builder()
                .With(PassingTests)
                .Deployed(false)
                .Build();

            _pipeline.Run(project);

            _log.LoggedLines
                .Should()
                .BeEquivalentTo("INFO: Tests passed",
                    "ERROR: Deployment failed",
                    "INFO: Sending email");

            _emailer.Received(1).Send("Deployment failed");
        }

        [Fact]
        public void Project_With_Tests_And_Failing_Build_Without_Email_Notification()
        {
            _config.SendEmailSummary().Returns(false);

            var project = Project.Builder()
                .With(PassingTests)
                .Deployed(false)
                .Build();

            _pipeline.Run(project);

            _log.LoggedLines
                .Should()
                .BeEquivalentTo("INFO: Tests passed",
                    "ERROR: Deployment failed",
                    "INFO: Email disabled");

            _emailer.DidNotReceive().Send(Any<string>());
        }

        [Fact]
        public void Project_Without_Tests_And_Failing_Build_With_Email_Notification()
        {
            _config.SendEmailSummary().Returns(true);

            var project = Project.Builder()
                .With(NoTests)
                .Deployed(false)
                .Build();

            _pipeline.Run(project);

            _log.LoggedLines
                .Should()
                .BeEquivalentTo("INFO: No tests",
                    "ERROR: Deployment failed",
                    "INFO: Sending email");

            _emailer.Received(1).Send("Deployment failed");
        }

        [Fact]
        public void Project_Without_Tests_And_Failing_Build_Without_Email_Notification()
        {
            _config.SendEmailSummary().Returns(false);

            var project = Project.Builder()
                .With(NoTests)
                .Deployed(false)
                .Build();

            _pipeline.Run(project);

            _log.LoggedLines
                .Should()
                .BeEquivalentTo("INFO: No tests",
                    "ERROR: Deployment failed",
                    "INFO: Email disabled");

            _emailer.DidNotReceive().Send(Any<string>());
        }
    }
}